pkg/controller/seed-controller-manager/initialmachinedeployment/controller.go

```diff
@@ -0,0 +1,243 @@
+		recorder:                      mgr.GetEventRecorderFor(ControllerName),
+		seedGetter:                    seedGetter,
+		userClusterConnectionProvider: userClusterConnectionProvider,
+		log:                           log,
+		versions:                      versions,
+	}
+
+	c, err := controller.New(ControllerName, mgr, controller.Options{
+		Reconciler:              reconciler,
+		MaxConcurrentReconciles: numWorkers,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create controller: %v", err)
+	}
+
+	if err := c.Watch(&source.Kind{Type: &kubermaticv1.Cluster{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("failed to create watch: %v", err)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	cluster := &kubermaticv1.Cluster{}
+	if err := r.Get(r.ctx, request.NamespacedName, cluster); err != nil {
+		if kerrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Add a wrapping here so we can emit an event on error
+	result, err := kubermaticv1helper.ClusterReconcileWrapper(
+		r.ctx,
+		r.Client,
+		r.workerName,
+		cluster,
+		r.versions,
+		kubermaticv1.ClusterConditionMachineDeploymentControllerReconcilingSuccess,
+		func() (*reconcile.Result, error) {
+			return r.reconcile(cluster)
+		},
+	)
+	if err != nil {
+		r.log.Errorw("Failed to reconcile cluster", "cluster", cluster.Name, zap.Error(err))
+		r.recorder.Event(cluster, corev1.EventTypeWarning, "ReconcilingError", err.Error())
+	}
+	if result == nil {
+		result = &reconcile.Result{}
+	}
+	return *result, err
+}
+
+func (r *Reconciler) reconcile(cluster *kubermaticv1.Cluster) (*reconcile.Result, error) {
+	// If cluster is not healthy yet there is nothing to do.
+	// If it gets healthy we'll get notified by the event. No need to requeue.
+	if !cluster.Status.ExtendedHealth.AllHealthy() {
+		return nil, nil
+	}
+
+	request := cluster.Annotations[v1.InitialMachineDeploymentRequestAnnotation]
+
+	// if there is no (more) annotation, we're done
+	if request == "" {
+		return nil, nil
+	}
+
+	var nodeDeployment *v1.NodeDeployment
+	if err := json.Unmarshal([]byte(request), &nodeDeployment); err != nil {
+		// TODO: What now? Remove the invalid annotation? This will end up in a reconcile loop.
+		return nil, fmt.Errorf("cannot unmarshal initial MachineDeployment request: %v", err)
+	}
+
+	nodeDeployment, err := machineresource.Validate(nodeDeployment, cluster.Spec.Version.Semver())
+	if err != nil {
+		// TODO: What now? Remove the invalid annotation? This will end up in a reconcile loop.
+		return nil, fmt.Errorf("initial node deployment is not valid: %v", err)
+	}
+
+	userClusterClient, err := r.userClusterConnectionProvider.GetClient(cluster)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get user cluster client: %v", err)
+	}
+
+	if err := r.createInitialMachineDeployment(nodeDeployment, cluster, userClusterClient); err != nil {
```

Comment:
It seems that you have removed the check for machine deployment existence here (previously in the `isInitialMachineDeploymentCreated` method). It was added to cover case in which initial machine deployment is created but then the annotation cannot be removed due some reason. If it happens now, we will have duplicated machine deployments or error (depends if the name was set in the JSON). It is not common thing to happen but perhaps we should cover it too, what do you think? However,`InitialMachineDeploymentLabel` label is additional side effect in my solution.